### ISSUE TITLE
Unable to load .hdf5 files

### ISSUE THREAD

**Issue by Miles-and-miles-away** [GliaLab/Begonia#9]
I am unable to load .hdf5 files using dataman. When I use Add file or Add directory it brings up the Locating data box, but it does not bring up anything. However, I was able to load a Tiff file. Am using Windows 10, Matlab 2021a.

**Comment by KnutAbjorsbraten** [comment]
Hi @123miles (: Could you tell us a little more about the .hdf5 file? How was it made, how you usually open it etc.? If you have an example file that fails to open, you can also send it to me?

**Comment by Miles-and-miles-away** [comment]
The HDF5 file is 4000 frames with 2 channels. In one channel is intracellular calcium and in the other is GRAB_NE noradrenaline biosensor. I thought ROA analysis might work for the biosensor signals. In the link below I have included a few file formats, although none were recognized by Dataman. When I try to load a file or folder containing these files, Dataman searches but then brings up nothing.

  1.  A single HDF5 with both channels
  2.  Two HDF5 files, one for each channel
  3.  Two HDF5 files, one for each channel in a format that is compatible with CaImAn.

https://filesender2.oist.jp/filesender/?s=download&token=9421d6bf-de26-4372-937a-1fe450d9ad8e

**Comment by Miles-and-miles-away** [comment]
Apologies, I didn’t answer your question. The files were made in Matlab (files attached)

**Comment by KnutAbjorsbraten** [comment]
A generous serving of bytes ;) I'm grabbing the files now and will discuss it with by colleague who implemented the HDF5 reader once I've had a peek.

**Comment by Miles-and-miles-away** [comment]
I tried downloading Begonia, and also changing the files to file.h5, but no luck.

**Comment by KnutAbjorsbraten** [comment]
Hi @milesOIST. We looked at the files. The files appear principally compatible.

The problem lies in loading metadata in a begonia-compatible way. We have not had time to implement a specific solution for that. The system is designed architecturally to allow loading of files from a variety of microscopes, so the primary way to add compatibility is to extend the `scantypes.TSeries` class for the microscope or software system used. In practice, we find HDF5 files vary in how they encode data and metadata.

An alternative is to use a set of default metadata for HDF5 files, but we suspect this would not be attractive for scientific work. Or maybe a generic metadata loader?

Is it within your reach to add compatibility for loading the files by extending the code? We could provide some technical assistance, e.g. a video chat to review what would be needed etc.
_
Vh, Knut

**Comment by Miles-and-miles-away** [comment]
Yes, I think we should be able to alter the output of our code. It would be great to talk it through. I'll email you to set up a time.

**Comment by KnutAbjorsbraten** [comment]
How is this going? (:

**Comment by phflot** [comment]
Hey everyboy, sorry for the late follow up - we managed to generate h5 files for Begonia and could successfully load them with Dataman. 
We are directly converting from MDF files which supply most of the metadata that Begonia stores in tseries objects. However, could you give me a short explanation of what's the difference between the parameters frame_count and slices and also what values I should supply to cycles, frame_position_um and source?
Small feature requests that could make it a bit easier for us:
- check also for '.HDF5' or '.HDF' extensions when reading files, or, alternatively, usage of the function H5F.is_hdf5(x) without restriction on extension to check whether any file is an hdf5 file
- add an attribute 'tseries' or 'begonia' ideally with version number that helps to identify hdf5 files that have been generated with Begonia

Best, Philipp

**Comment by Miles-and-miles-away** [comment]
I have managed to get the H5 files produced by @phflot 's Flow Registration package loaded into Dataman. Stabilization was still failing, but after reinstalling Begonia it now works. I've also managed to label ROIs and find ROAs. 

I found one other issue which occurs for the files created by exporting TIFF files from MDF files using MView.. When I tried adding TIFF files directly into Dataman, I am able to convert to H5 and Stabilize, but when I try to play the file in `Mark ROIs` the file I get this error:
```
Error while evaluating TimerFcn for timer 'timer-2'
Unrecognized field name "autoplay_start_time".
```

**Comment by Knutsi** [comment]
Hi @milesOIST

> I found one other issue which occurs for the files created by exporting TIFF files from MDF files using MView.. When I tried adding TIFF files directly into Dataman, I am able to convert to H5 and Stabilize, but when I try to play the file in `Mark ROIs` the file I get this error:
> 
> ```
> Error while evaluating TimerFcn for timer 'timer-2'
> Unrecognized field name "autoplay_start_time".
> ```


The error is very difficult unfortunately. The only way I managed to get roiman to render graphics effectively was to use timers, and Matlab is very bad at informing about what goes wrong inside these timers. Grabbing your file now to have a looksee.

> Hey everyboy, sorry for the late follow up - we managed to generate h5 files for Begonia and could successfully load them with Dataman. We are directly converting from MDF files which supply most of the metadata that Begonia stores in tseries objects. However, could you give me a short explanation of what's the difference between the parameters frame_count and slices and also what values I should supply to cycles, frame_position_um and source? 

@phflot : 
- not sure about slices - maybe @lastis  can answer.
- cycles is in support of Bruker microscopes that can repeat sets of recordings. In practice we don't use this any code in the public repo I think
- frame_position_um is the position of the imaging plane center coordinate in 3D space. We use this to map rois, planes and points inside z-stacks recordings
- source indicates the type of microscope e.g. that recorded the image series

**Comment by lastis** [comment]
Hey Phillip! @phflot 
> Hey everyboy, sorry for the late follow up - we managed to generate h5 files for Begonia and could successfully load them with Dataman. We are directly converting from MDF files which supply most of the metadata that Begonia stores in tseries objects. However, could you give me a short explanation of what's the difference between the parameters frame_count and slices and also what values I should supply to cycles, frame_position_um and source? Small feature requests that could make it a bit easier for us:

Regarding frames vs. slices. The parameter 'slices' in the metadata struct is not used and was only added because we might use the number of slices in the ZStack class. Just to be clear, I use the term 'slices' to represent the number of images in the z-dimension. The TSeries class is only meant to represent 3D image stacks (x,y,t), and the ZStack class is meant to represent 4D image stacks (x,y,z,time). We/I have used the ZStack class very little and I do not think it is actually functioning. For the TSeries class you only need to supply the frame_count. As you probably are aware by now the imaging code is in a continuous state of 'work-in-progress' 😛

On a side note, in the TSeriesTiff class when reading a tiff file from FIJI the number of slices are assigned to the metadata.frame_count if the number of frames is missing. This was just to have a little broader support for reading tiffs. 

> * check also for '.HDF5' or '.HDF' extensions when reading files, or, alternatively, usage of the function H5F.is_hdf5(x) without restriction on extension to check whether any file is an hdf5 file
> * add an attribute 'tseries' or 'begonia' ideally with version number that helps to identify hdf5 files that have been generated with Begonia

I think this is a great idea! I'm not sure when/how I will implement this yet. If you would like to give a code suggestion I am very happy to test it :)

**Comment by Miles-and-miles-away** [comment]
@Knutsi I have found that the timer problem occurs only when I try to play the video. If I press the Play button on the Playback window, it freezes. However, if I open the window and then manually drag the slider, it works. This occurs both with marking the ROIs as well as with checking the ROA mask. It seems it is an issue with Play function. It also happens with both the stable and unstable versions.
